src/main/java/core/basesyntax/MyLinkedList.java

```diff
@@ -3,48 +3,152 @@
 import java.util.List;
 
 public class MyLinkedList<T> implements MyLinkedListInterface<T> {
+    private Node<T> head;
+    private Node<T> tail;
+    private int size;
+
+    public MyLinkedList() {
+        this.head = null;
+        this.tail = null;
+        this.size = 0;
+    }
+
+    private static class Node<T> {
+        private Node<T> prev;
+        private T value;
+        private Node<T> next;
+
+        public Node(Node<T> prev, T value, Node<T> next) {
+            this.prev = prev;
+            this.value = value;
+            this.next = next;
+        }
+    }
+
+    private boolean isInvalidIndex(int index) {
+        return index < 0 || index >= size;
+    }
+
+    private void linkToNode(T elementToInsert, Node<T> successor) {
+        Node<T> predecessor = successor.prev;
+        Node<T> newNode = new Node<>(predecessor, elementToInsert, successor);
+        successor.prev = newNode;
+        if (predecessor == null) {
+            head = newNode;
+        } else {
+            predecessor.next = newNode;
+        }
+        size++;
+    }
+
+    private void linkToTail(T elementToInsert) {
+        Node<T> predecessor = this.tail;
+        Node<T> newNode = new Node<>(predecessor, elementToInsert, null);
+        this.tail = newNode;
+        if (predecessor == null) {
+            head = newNode;
+        } else {
+            predecessor.next = newNode;
+        }
+        size++;
+    }
+
+    private Node<T> getNodeByIndex(int index) {
+        Node<T> targetNode = head;
+        for (int i = 0; i < index; i++) {
+            targetNode = targetNode.next;
+        }
+        return targetNode;
+    }
+
+    private T removeNode(Node<T> nodeToRemove) {
+        Node<T> beforeTarget = nodeToRemove.prev;
+        Node<T> afterTarget = nodeToRemove.next;
+        if (afterTarget == null) {
+            tail = beforeTarget;
+            nodeToRemove.prev = null;
+        } else if (beforeTarget == null) {
+            head = afterTarget;
+            nodeToRemove.next = null;
+        } else {
+            beforeTarget.next = afterTarget;
+            afterTarget.prev = beforeTarget;
+        }
+        size--;
+        return nodeToRemove.value;
+    }
+
     @Override
     public boolean add(T value) {
-        return false;
+        linkToTail(value);
+        return true;
     }
 
     @Override
     public void add(T value, int index) {
-
+        if (index == size) {
+            linkToTail(value);
+        } else if (index < size && index >= 0) {
+            linkToNode(value, getNodeByIndex(index));
+        } else {
```

Comment:
move throwing of exception into if block, so you won't have additional else clause

Reply:
done